src/build_loop.rs

```diff
@@ -63,43 +63,65 @@ impl<'a> BuildLoop<'a> {
     /// Sends `Event`s over `Self.tx` once they happen.
     /// When new filesystem changes are detected while a build is
     /// still running, it is finished first before starting a new build.
-    pub fn forever(&mut self, tx: chan::Sender<Event>) {
+    #[allow(clippy::drop_copy, clippy::zero_ptr)] // triggered by `select!`
+    pub fn forever(&mut self, tx: chan::Sender<Event>, rx_ping: chan::Receiver<()>) {
         let send = |msg| tx.send(msg).expect("Failed to send an event");
+        let translate_reason = |rsn| match rsn {
+            Ok(rsn) => rsn,
+            // we should continue and just cite an unknown reason
+            Err(EventError::EventHasNoFilePath(msg)) => {
+                warn!(
+                    "Event has no file path; possible issue with the watcher?: {:#?}",
+                    msg
+                );
+                // can’t Clone `Event`s, so we return the Debug output here
+                Reason::UnknownEvent(DebugMessage::from(format!("{:#?}", msg)))
+            }
+            Err(EventError::RxNoEventReceived) => {
+                panic!("The file watcher died!");
+            }
+        };
 
-        send(Event::Started(Reason::ProjectAdded(
+        // The project has just been added, so run the builder in the first iteration
+        let mut reason = Some(Event::Started(Reason::ProjectAdded(
             self.project.nix_file.clone(),
         )));
+        let mut output_paths = None;
+
+        // Drain pings initially: we're going to trigger a first build anyway
+        rx_ping.try_iter().for_each(drop);
```

Comment:
Is it possible this will drain pings for other projects, too?

Reply:
No: `rx_ping` is created by `Daemon::add`, so there is one channel per project. For each project, there is a `Handler` in `Daemon::handler_threads` which contains the sender `tx` corresponding to the `rx_ping` receiver. It's somewhat implicit, but I hope that this structure makes it clear that it is _intended_ to have one of those channels per project.